core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala

```diff
@@ -329,23 +329,62 @@ trait AttributeComponent {
       recs.map { rec => (AttributeRecordPrimaryKey(rec.ownerId, rec.namespace, rec.name, rec.listIndex), rec) }.toMap
 
     def patchAttributesAction(inserts: Traversable[RECORD], updates: Traversable[RECORD], deleteIds: Traversable[Long], insertFunction: Seq[RECORD] => String => WriteAction[Int]) = {
-      deleteAttributeRecordsById(deleteIds.toSeq) andThen
-        batchInsertAttributes(inserts.toSeq) andThen
-        AlterAttributesUsingScratchTableQueries.updateAction(insertFunction(updates.toSeq))
+      for {
+        _ <- if (deleteIds.nonEmpty) deleteAttributeRecordsById(deleteIds.toSeq) else DBIO.successful(0)
+        _ <- if (inserts.nonEmpty) batchInsertAttributes(inserts.toSeq) else DBIO.successful(0)
+        updateResult <- if (updates.nonEmpty)
+                          AlterAttributesUsingScratchTableQueries.updateAction(insertFunction(updates.toSeq))
+                        else
+                          DBIO.successful(0)
+      } yield {
+        updateResult
+      }
     }
 
     def rewriteAttrsAction(attributesToSave: Traversable[RECORD], existingAttributes: Traversable[RECORD], insertFunction: Seq[RECORD] => String => WriteAction[Int]) = {
       val toSaveAttrMap = toPrimaryKeyMap(attributesToSave)
       val existingAttrMap = toPrimaryKeyMap(existingAttributes)
 
-      // update attributes which are in both to-save and currently-exists, insert attributes which are in save but not exists
-      val (attrsToUpdateMap, attrsToInsertMap) = toSaveAttrMap.partition { case (k, v) => existingAttrMap.keySet.contains(k) }
-      // delete attributes which currently exist but are not in the attributes to save
-      val attributesToDelete = existingAttrMap.filterKeys(! attrsToUpdateMap.keySet.contains(_)).values
+      val existingKeys = existingAttrMap.keySet
+
+      // insert attributes which are in save but not exists
+      val attributesToInsert = toSaveAttrMap.filterKeys(! existingKeys.contains(_))
+
+      // delete attributes which are in exists but not save
+      val attributesToDelete = existingAttrMap.filterKeys(! toSaveAttrMap.keySet.contains(_))
+
+      // update attributes which are in both to-save and currently-exists, but have different values.
+      // note that currently-existing attributes will have a populated id e.g. "1234", but to-save will have an id of "0"
+      // therefore, use a comparison function that looks at everything except id
+      // note this does not compare transactionId for AttributeScratchRecords. We do not expect AttributeScratchRecords
+      // here, and transactionId will eventually be going away, so don't bother
+      // TODO: should this compare function move closer to the AttributeRecord class?
+      def equalRecords(left: RECORD, right: RECORD): Boolean = {
+        // compare everything except id
```

Comment:
i was wondering why you hadn't done something like `left.copy(id=0).equals(right.copy(id=0))` but i see now that switching this to take an `AttributeRecord` to do that would probably be annoying and is _not_ worth the effort right now